text/3735-field-projections.md

```diff
@@ -0,0 +1,1744 @@
+this section.
+
+Structurally pinned fields are marked with `#[pin]` using the derive macro `PinProject`. For example
+consider a future that alternatingly polls two futures:
+
+```rust
+#[derive(PinProject)]
+struct FairRaceFuture<F1, F2> {
+    #[pin]
+    fut1: F1,
+    #[pin]
+    fut2: F2,
+    fair: bool,
+}
+```
+
+Now, it's possible to project a `fut: Pin<&mut FairRaceFuture>`:
+- `fut->fut1: Pin<&mut F1>`
+- `fut->fut2: Pin<&mut F2>`
+- `fut->fair: &mut bool`
+
+Using these, one can concisely implement `Future` for `FairRaceFuture` without any `unsafe` code:
+
+```rust
+impl<F1: Future, F2: Future<Output = F1::Output>> Future for FairRaceFuture<F1, F2> {
+    type Output = F1::Output;
+
+    fn poll(self: Pin<&mut Self>, ctx: &mut Context) -> Poll<Self::Output> {
+        let fair: &mut bool = self->fair;
+        *fair = !*fair;
+        if *fair {
+            // self->fut1: Pin<&mut F1>
+            match self->fut1.poll(ctx) {
+                Poll::Pending => self->fut2.poll(ctx),
+                Poll::Ready(res) => Poll::Ready(res),
+            }
+        } else {
+            // self->fut2: Pin<&mut F2>
+            match self->fut2.poll(ctx) {
+                Poll::Pending => self->fut1.poll(ctx),
+                Poll::Ready(res) => Poll::Ready(res),
+            }
+        }
+    }
+}
+```
+
+### Implementing Custom Field Projections
+
+There are two different ways to implement projections for a custom type:
+- implementing the `Projectable` and `Project<F>` traits, or
+- annotating it with `#[projecting]`.
+
+They are used for pointers and container types respectively.
+
+#### Pointer-Like
+
+Pointer-like types can be projected using `Projectable` and `Project<F>`. For example, if we create
+a custom reference type that simultaneously points at two instances of the same type.
+
+```rust
+pub struct DoubleRef<'a, T> {
+    first: &'a T,
+    second: &'a T,
+}
+
+impl<'a, T> DoubleRef<'a, T> {
+    pub fn new(first: &'a T, second: &'a T) -> Self {
+        Self { first, second }
+    }
+
+    pub fn read(&self) -> (T, T)
+    where
+        T: Copy
+    {
+        (*self.first, *self.second)
+    }
+}
+```
+
+We can now allow its users to use field projection by implementing the above mentioned traits:
+
+```rust
+impl<'a, T> Projectable for DoubleRef<'a, T> {
+    type Inner = T;
+}
+```
+
+The `Projectable` trait is what tells the compiler which types' fields to consider for projecting.
+When you write `a->b`, then `a` has to implement `Projectable` in order for the compiler to know
+which type to look up the field `b` for.
+
+The actual projection operation is governed by `Project<F>`:
+
+```rust
+impl<'a, T, F> Project<F> for DoubleRef<'a, T>
+where
+    F: Field<Base = T>,
+{
+    type Output = DoubleRef<'a, F::Type>;
+
+    fn project(self) -> Self::Output {
+        DoubleRef {
+            first: self.first.project(),
+            second: self.second.project(),
```

Comment:
How does this work, how does it know which field to project them to?
Is this `Project::<F>::project(self.first)`? (Not sure if there's a nicer way to fix the `F` here. Maybe `self.first->F` should work. ;)

Reply:
I expect this works the same way you can just call `self.my_field.fmt(f)` in a `Debug` impl even though that's not the only trait with a `fmt` function that may be in scope -- it's the only impl in the `where` bounds for the generic arguments.

Reply:
Exactly, I was just hoping that the type inference would be good enough to achieve this, I didn't test it though.